File: strategy/src/strategy.rs
#![allow(unused_variables)]

use std::{collections::HashMap, sync::Arc};

use async_trait::async_trait;
use chrono::{Local, NaiveDate, NaiveDateTime, NaiveTime};
use hiq_data::{store::Loader, Bar};
use serde::{Deserialize, Serialize};

use crate::{Result, StrategyType};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Stat {
    pub hit_chg_pct: [f32; 5], // 1, 2, 4, 8, now 涨幅
    pub start: NaiveDate,
    pub end: NaiveDate,
    pub low: NaiveDate,
    pub high: NaiveDate,
    pub hit: NaiveDate,
    pub hit_max: NaiveDate,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct StrategyResult {
    pub code: String,
    pub name: String,
    pub mark: Option<HashMap<NaiveDate, String>>,
    pub stat: Option<Stat>,
}

impl StrategyResult {
    pub fn new(
        code: String,
        name: String,
        mark: Option<HashMap<NaiveDate, String>>,
        stat: Option<Stat>,
    ) -> Self {
        Self {
            code,
            name,
            mark,
            stat,
        }
    }
}

#[derive(Clone, Debug)]
pub struct CommonParam {
    pub test_end_date: Option<NaiveDateTime>,
    pub test_trade_days: Option<i64>,
}

impl CommonParam {
    pub fn new(test_end_date: Option<NaiveDateTime>, test_trade_days: Option<i64>) -> Self {
        Self {
            test_end_date,
            test_trade_days,
        }
    }
}

impl Default for CommonParam {
    fn default() -> Self {
        let n = Local::now().date_naive();
        let n = NaiveDateTime::new(n, NaiveTime::from_hms_opt(0, 0, 0).unwrap());
        Self {
            test_end_date: Some(n),
            test_trade_days: Some(60)
        }
    }
}

#[async_trait]
pub trait Strategy: Sync + Send {
    fn help(&self) -> String {
        String::from("")
    }
    fn name(&self) -> String {
        String::from("")
    }
    async fn prepare(
        &mut self,
        loader: Arc<Box<dyn Loader>>,
        cmm_params: Option<CommonParam>,
        params: Option<HashMap<String, String>>,
    ) -> Result<()> {
        Ok(())
    }
    fn accept(&self) -> Vec<StrategyType> {
        vec![StrategyType::Stock]
    }
    async fn test(
        &self,
        loader: Arc<Box<dyn Loader>>,
        typ: StrategyType,
        code: String,
        name: String,
    ) -> Result<Option<StrategyResult>> {
        Ok(None)
    }
}

pub fn stat_result(data: &Vec<Bar>, hit: usize, hit_max: usize) -> Result<Stat> {
    if hit >= data.len() || hit_max > data.len() || hit_max < hit {
        return Err(crate::Error::Custom(format!(
            "hit={} or hit_max={}, out of range(0~{}) or invalid",
            hit,
            hit_max,
            data.len()
        )));
    }
    let hit_bar = data
        .get(hit)
        .ok_or(crate::Error::Custom("invalid index".to_string()))?;
    let hit_max_bar = data
        .get(hit_max)
        .ok_or(crate::Error::Custom("invalid index".to_string()))?;
    let start_bar = data
        .last()
        .ok_or(crate::Error::Custom("invalid index".to_string()))?;
    let end_bar = data
        .first()
        .ok_or(crate::Error::Custom("invalid index".to_string()))?;

    let (mut low, mut high) = (0.0, 0.0);
    let (mut low_index, mut high_index) = (0usize, 0usize);
    for (i, val) in data.iter().enumerate() {
        if val.close < low {
            low = val.close;
            low_index = i;
        }
        if val.close > high {
            high = val.close;
            high_index = i;
        }
    }
    let low_bar = data
        .get(low_index)
        .ok_or(crate::Error::Custom("invalid index".to_string()))?;
    let high_bar = data
        .get(high_index)
        .ok_or(crate::Error::Custom("invalid index".to_string()))?;

    let mut hit_chg_pct: [f32; 5] = [0.0; 5];
    let mut hit_chg_pct_index = 0;
    for i in 0..hit {
        let index = hit - i - 1;
        let step = i + 1;
        if step == 1 || step == 2 || step == 4 || step == 8 {
            let bar = data
                .get(index)
                .ok_or(crate::Error::Custom("invalid index".to_string()))?;
            hit_chg_pct[hit_chg_pct_index] = (bar.close - hit_bar.close) * 100.0 / hit_bar.close;
            hit_chg_pct_index += 1;
        }
    }
    let now_bar = data
        .get(0)
        .ok_or(crate::Error::Custom("invalid index".to_string()))?;
    hit_chg_pct[4] = (now_bar.close - hit_bar.close) * 100.0 / hit_bar.close;
    let stat = Stat {
        hit_chg_pct, // 1, 2, 4, 8, now 涨幅
        start: start_bar.trade_date.date(),
        end: end_bar.trade_date.date(),
        low: low_bar.trade_date.date(),
        high: high_bar.trade_date.date(),
        hit: hit_bar.trade_date.date(),
        hit_max: hit_max_bar.trade_date.date(),
    };
    Ok(stat)
}

pub type ProgressFunc = Box<dyn Fn(&str, &str, usize, usize, f32) -> () + Sync + Send>;

File: python/hiq-pystrategy/hiq_pystrategy/strategy.py
from abc import ABC
from datetime import date
import json
from typing import Dict, List, Optional
from hiq_pydata.hiq_loader import HiqLoader
from hiq_pyfetch.hiq_fetch import HiqFetch
from hiq_pystrategy.hiq_pystrategy import stat_result

def _json_def_handler(obj):
    if hasattr(obj, 'isoformat'):
        return obj.isoformat()
    return None

class StrategyType:
    Bond = 1
    Fund = 2
    Stock = 3
    Index = 4
    Concept = 5
    Industry = 6

    @staticmethod
    def str(typ) -> str:
        m = {
            StrategyType.Bond: 'Bond',
            StrategyType.Fund: 'Fund',
            StrategyType.Stock: 'Stock',
            StrategyType.Index: 'Index',
            StrategyType.Concept: 'Concept',
            StrategyType.Industry: 'Industry'
        }
        if typ in m.keys():
            return m[typ]
        return ''


class Stat:
    def __init__(self, hit_chg_pct=None, start=None, end=None, low=None, high=None, hit=None, hit_max=None) -> None:
        self.hit_chg_pct = hit_chg_pct
        self.start = start
        self.end = end
        self.low = low
        self.high = high
        self.hit = hit
        self.hit_max = hit_max

    def to_dict(self):
        return dict(hit_chg_pct=self.hit_chg_pct,
                    start=self.start,
                    end=self.end,
                    low=self.low,
                    high=self.high,
                    hit=self.hit,
                    hit_max=self.hit_max)

    def from_json(self, js):
        self.hit_chg_pct = js['hit_chg_pct']
        self.start = js['start']
        self.end = js['end']
        self.low = js['low']
        self.high = js['high']
        self.hit = js['hit']
        self.hit_max = js['hit_max']

    def to_json(self):
        return json.dumps(self.to_dict(), default=_json_def_handler)


class StrategyResult:
    def __init__(self, code=None, name=None, mark=None, stat=None) -> None:
        self.code = code
        self.name = name
        self.mark = mark
        self.stat = stat

    def to_dict(self):
        return dict(code=self.code, name=self.name, stat=self.stat.to_dict() if self.stat is not None else None)

    def from_json(self, js):
        self.code = js['code']
        self.name = js['name']
        if 'mark' in js:
            self.mark = js['mark']

        if 'stat' in js:
            s = Stat()
            s.from_json(js['stat'])
            self.stat = s

    def to_json(self):
        return json.dumps(self.to_dict(), default=_json_def_handler)


class CommonParam:
    def __init__(self, test_end_date, test_trade_days) -> None:
        self.test_end_date = test_end_date
        self.test_trade_days = test_trade_days


class Strategy(ABC):
    def __init__(self, loader: Optional[HiqLoader] = None, fetch: Optional[HiqFetch] = None, cmm_params: Optional[CommonParam] = None, params: Optional[Dict] = None) -> None:
        self.loader = loader
        self.fetch = fetch
        self.cmm_params = cmm_params
        self.params = params

    async def run(self, typ, code, name) -> Optional[str]:
        """
        由runner调用，不要重写
        """
        rs = await self.test(typ, code, name)
        if rs is not None:
            rs = rs.to_json()
        return rs

    @staticmethod
    def stat_result(data: List[Dict], hit: int, hit_max: int) -> Stat:
        js_str = json.dumps(data, default=_json_def_handler)
        js_str = stat_result(js_str, hit, hit_max)
        s = Stat()
        s.from_json(js_str)
        return s

    def help(self) -> str:
        return ""

    def name(self) -> str:
        return self.__qualname__.name

    def accept(self) -> List[int]:
        return [StrategyType.Stock]

    async def prepare(self) -> bool:
        return True

    async def test(self, typ: StrategyType, code: str, name: str) -> Optional[StrategyResult]:
        pass
